lsa: decay reserves

avikivity opened the pull request:
The log-structured allocator (LSA) reserves memory when performing
operations, since its operations are performed with reclaiming disabled
and if it runs out, it cannot evict cache to gain more. The amount of
memory to reserve is remembered across calls so that it does not have
to repeat the fail/increase-reserve/retry cycle for every operation.

However, we currently lack decaying the amount to reserve. This means
that if a single operation increased the reserve in the distant past,
all current operations also require this large reserve. Large reserves
are expensive since they can cause large amounts of cache to be evicted.

This patch adds reserve decay. The time-to-decay is inversely proportional
to reserve size: 10GB/reserve. This means that a 20MB reserve is halved
after 500 operations (10GB/20MB) while a 20kB reserve is halved after
500,000 operations (10GB/20kB). So large, expensive reserves are decayed
quickly while small, inexpensive reserves are decayed slowly to reduce
the risk of allocation failures and exceptions.

Fixes #325.

tgrabiec commented on utils/logalloc.hh:
```diff
@@ -685,16 +685,24 @@ struct reclaim_lock {
-    size_t _lsa_reserve = 10; // in segments
-    size_t _std_reserve = 1024; // in bytes
+    // Do not decay below these minimal values
+    static constexpr size_t s_min_lsa_reserve = 10;
```
Is there some new convention behind the `s_` name prefix (s = static)?

avikivity replied:
Yes, not formalized.

avikivity replied:
(and not new, see s_shard_id, s_rpc_client_idx_table)

tgrabiec commented on utils/logalloc.cc:
```diff
@@ -2259,7 +2259,36 @@ allocating_section::guard::~guard() {
+    // decay it at a lower rate.
+
+    _remaining_lsa_segments_until_decay -= _lsa_reserve;
+    if (_remaining_lsa_segments_until_decay < 0) {
```
_remaining_lsa_segments_until_decay is unsigned, how can it go below 0?

avikivity replied:
Oops, meant to change it to be signed, but forgot.

tgrabiec commented on utils/logalloc.cc:
```diff
@@ -2259,7 +2259,36 @@ allocating_section::guard::~guard() {
+
+    _remaining_std_bytes_until_decay -= _std_reserve;
+    if (_remaining_std_bytes_until_decay < 0) {
+        _remaining_std_bytes_until_decay = s_segments_per_decay;
```
Should be `s_bytes_per_decay`

avikivity replied:
Yes.

tgrabiec commented on utils/logalloc.cc:
```diff
@@ -2259,7 +2259,36 @@ allocating_section::guard::~guard() {
+}
+
 void allocating_section::reserve() {
+  maybe_decay_reserve();
```
with_reclaiming_disabled() calls reserve() when retrying, in which case we should not attempt to decay, otherwise they may cancel each other and in the extreme case we'll not progress.

I suggest to call it once per with_reclaiming_disabled.

avikivity replied:
You mean, if the number of attempts is equal to the number of times until decay? I think it's unlikely, but will change to reduce the risk of confusion.

tgrabiec commented on utils/logalloc.cc:
```diff
@@ -2259,7 +2259,36 @@ allocating_section::guard::~guard() {
+void allocating_section::maybe_decay_reserve() {
+    // The decay rate is inversely proportional to the reserve
+    // (every (s_segments_per_decay/_lsa_reserve) allocations).
+    //
```
Approach LGTM

avikivity replied:
Thanks. I'll try to add a unit test for the next iteration.

avikivity commented:
Update: 

 - changed _remaining_std_bytes_until_decay, _remaining_lsa_segments_until_decay to be signed, so that the checks that they are < 0 make sense
 - decay std reserve by the correct constant, not s_segments_per_decay
 - moved decay to with_reclaiming_disabled() to prevent chance of decay happening more than once per operation
 - removed FIXME: decay comments
 - add unit test

avikivity commented:
Queued.

avikivity commented:
Merged as 7ac59dcc9837658bb9fc1ff06dc4c95fdabef6f8
